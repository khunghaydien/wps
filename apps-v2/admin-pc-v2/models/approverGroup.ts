type Department = {
  code: string;
  name: string;
};

export type Approver = {
  id?: string;
  username?: string;
};

export type ApproverGroup = {
  id: string;
  companyId: string;
  code: string;
  department: Department;
  name: string;
  name_L0: string;
  name_L1: string;
  name_L2: string;
  queueId: string;
  attQueueId: string;
  expQueueId: string;
  generalQueueId: string;
  timeQueueId: string;
  type: keyof typeof TYPE_LABELS;
  targetDepartmentId: string;
  lastUpdatedAt: string;
};

export const TYPE_LABELS = {
  DeptManagers: 'Admin_Lbl_DepartmentManagers',
  DeptMembers: 'Admin_Lbl_DepartmentMembers',
  SpecificUsers: 'Admin_Lbl_SpecificUsers',
};

export const APPROVER_MODULES = {
  attQueueId: 'Admin_Lbl_EmpSectAttendance',
  expQueueId: 'Admin_Lbl_EmpSectExpense',
  timeQueueId: 'Admin_Lbl_EmpSectTimeTracking',
  generalQueueId: 'Admin_Lbl_CustomRequest',
};
