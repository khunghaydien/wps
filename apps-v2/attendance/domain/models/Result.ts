export type ResultSuccess<Value = void> = { result: true; value: Value };

export type ResultFailed<Reason> = {
  result: false;
  reason: Reason;
};

export type Result<Reason, Value = unknown> =
  | ResultSuccess<Value>
  | ResultFailed<Reason>;

export const REASON = {
  UNEXPECTED: 'unexpected',
  USER_INDUCED: 'userInduced',
  NO_RECORD: 'noRecord',
  EXISTED_INVALID_REQUEST: 'existedInvalidRequest',
  EXISTED_SUBMITTING_REQUEST: 'existedSubmittingRequest',
} as const;

export type ReasonMap = typeof REASON;

export type UnexpectedReason = ReasonMap['UNEXPECTED'];

export type UserInducedReason = ReasonMap['USER_INDUCED'];

export type NoRecord = ReasonMap['NO_RECORD'];

export type NotSubmittedReason =
  | ReasonMap['EXISTED_INVALID_REQUEST']
  | ReasonMap['EXISTED_SUBMITTING_REQUEST'];
