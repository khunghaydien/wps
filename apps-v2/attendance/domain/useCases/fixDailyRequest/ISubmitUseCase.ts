import { Status as DailyRequestStatus } from '@attendance/domain/models/AttDailyRequest';
import {
  NotSubmittedReason,
  Result,
  UserInducedReason,
} from '@attendance/domain/models/Result';

import * as Interface from '@attendance/domain/useCases/IUseCase';

export type IInputData = Interface.IInputData<{
  id: string;
  dailyRequestSummary: {
    status: DailyRequestStatus;
  };
}>;

export type IOutputData = Interface.IOutputData<
  Result<UserInducedReason | NotSubmittedReason>
>;

export type IUseCase = Interface.IUseCase<IInputData, IOutputData>;
