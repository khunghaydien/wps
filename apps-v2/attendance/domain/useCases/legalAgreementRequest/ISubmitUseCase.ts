import { Code } from '@attendance/domain/models/LegalAgreementRequestType';

import * as Interface from '@attendance/domain/useCases/IUseCase';

export type IInputData = Interface.IInputData<{
  requestId: string | null;
  summaryId: string;
  requestType: Code;
  changedOvertimeHoursLimit: number;
  reason: string;
  measures: string;
}>;

export type IOutputData = Interface.IOutputData<void>;

export type IUseCase = Interface.IUseCase<IInputData, IOutputData>;
