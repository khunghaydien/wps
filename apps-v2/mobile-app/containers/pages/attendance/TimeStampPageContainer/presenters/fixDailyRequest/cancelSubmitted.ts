import { Store } from 'redux';

import msg from '@apps/commons/languages';
import { showToast } from '@commons/modules/toast';
import { AppDispatch } from '@mobile/modules/commons/AppThunk';
import { showConfirm } from '@mobile/modules/commons/confirm';
import { catchApiError } from '@mobile/modules/commons/error';
import { endLoading, startLoading } from '@mobile/modules/commons/loading';

import { actions as uiActions } from '@mobile/modules/attendance/timeStamp/ui';

import { IPresenter } from '@attendance/application/useCaseInteractors/fixDailyRequest/CancelApprovalUseCaseInteractor';

export default ({ dispatch }: Store) =>
  (): IPresenter => {
    let loadingId;
    const $dispatch = dispatch as AppDispatch;
    return {
      start: () => {
        loadingId = $dispatch(startLoading());
      },
      confirm: async () => {
        $dispatch(endLoading(loadingId));
        const result = await $dispatch(
          showConfirm(msg().Appr_Msg_RequestConfirmCancelRequest)
        );
        loadingId = $dispatch(startLoading());
        return result;
      },
      complete: ({ result }) => {
        if (result) {
          $dispatch(showToast(msg().Att_Msg_CanceledFixDailyRequest));
          $dispatch(uiActions.clearComment());
        }
      },
      error: (err) => {
        $dispatch(catchApiError(err as Parameters<typeof catchApiError>[0]));
      },
      finally: () => {
        $dispatch(endLoading(loadingId));
      },
    };
  };
