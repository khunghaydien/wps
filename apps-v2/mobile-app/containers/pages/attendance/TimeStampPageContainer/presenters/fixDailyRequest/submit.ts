import { Store } from 'redux';

import msg from '@apps/commons/languages';
import { showToast } from '@commons/modules/toast';
import { AppDispatch } from '@mobile/modules/commons/AppThunk';
import {
  catchApiError,
  catchBusinessError,
} from '@mobile/modules/commons/error';
import { endLoading, startLoading } from '@mobile/modules/commons/loading';

import { REASON } from '@attendance/domain/models/Result';

import { State } from '@mobile/modules';
import { actions as uiActions } from '@mobile/modules/attendance/timeStamp/ui';

import askIgnoreWarning from '@mobile/action-dispatchers/attendance/userPrompts/askAttendanceRequestIgnoreWarning';

import { IPresenter } from '@attendance/application/useCaseInteractors/fixDailyRequest/SubmitUseCaseInteractor';

export default (store: Store) => (): IPresenter => {
  let loadingId;
  const $dispatch = store.dispatch as AppDispatch;
  return {
    start: () => {
      loadingId = $dispatch(startLoading());
    },
    confirmToSubmitWithWarning: async (confirmation) => {
      $dispatch(endLoading(loadingId));

      const result = await $dispatch(askIgnoreWarning(confirmation));
      loadingId = $dispatch(startLoading());
      return result;
    },
    complete: (output) => {
      if (output.result === true) {
        const state = store.getState() as State;
        if (!state?.attendance?.timeStamp?.ui?.sending) {
          $dispatch(showToast(msg().Att_Msg_FixDailyRequestIsRequested));
        }
        $dispatch(uiActions.clearComment());
        return;
      }
      switch (output.reason) {
        case REASON.EXISTED_INVALID_REQUEST:
          $dispatch(
            catchBusinessError(
              msg().Com_Lbl_Error,
              msg().Att_Err_InvalidRequestExist,
              msg().Att_Slt_InvalidRequestExist
            )
          );
          return;
        case REASON.EXISTED_SUBMITTING_REQUEST:
          $dispatch(
            catchBusinessError(
              msg().Com_Lbl_Error,
              msg().Att_Err_RequestingRequestExist,
              msg().Att_Slt_RequestingRequestExist
            )
          );
      }
    },
    error: (err) => {
      $dispatch(catchApiError(err as Parameters<typeof catchApiError>[0]));
    },
    finally: () => {
      $dispatch(endLoading(loadingId));
    },
  };
};
