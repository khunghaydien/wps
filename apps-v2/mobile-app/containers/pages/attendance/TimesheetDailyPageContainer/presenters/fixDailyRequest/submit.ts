import { Store } from 'redux';

import msg from '@apps/commons/languages';
import { showToast } from '@commons/modules/toast';
import { AppDispatch } from '@mobile/modules/commons/AppThunk';
import {
  catchApiError,
  catchBusinessError,
} from '@mobile/modules/commons/error';
import { endLoading, startLoading } from '@mobile/modules/commons/loading';

import askIgnoreWarning from '@mobile/action-dispatchers/attendance/userPrompts/askAttendanceRequestIgnoreWarning';

import { IPresenter } from '@attendance/application/useCaseInteractors/fixDailyRequest/SubmitUseCaseInteractor';

export default ({ dispatch }: Store) =>
  (): IPresenter => {
    let loadingId;
    const $dispatch = dispatch as AppDispatch;
    return {
      start: () => {
        loadingId = $dispatch(startLoading());
      },
      confirmToSubmitWithWarning: async (confirmation) => {
        $dispatch(endLoading(loadingId));

        const result = await $dispatch(askIgnoreWarning(confirmation));
        loadingId = $dispatch(startLoading());
        return result;
      },
      complete: (output) => {
        if (output.result === true) {
          $dispatch(showToast(msg().Att_Msg_FixDailyRequestIsRequested));
          return;
        }
        switch (output.reason) {
          case 'existedInvalidRequest':
            dispatch(
              catchBusinessError(
                msg().Com_Lbl_Error,
                msg().Att_Err_InvalidRequestExist,
                msg().Att_Slt_InvalidRequestExist
              )
            );
            return;
          case 'existedSubmittingRequest':
            dispatch(
              catchBusinessError(
                msg().Com_Lbl_Error,
                msg().Att_Err_RequestingRequestExist,
                msg().Att_Slt_RequestingRequestExist
              )
            );
        }
      },
      error: (err) => {
        dispatch(catchApiError(err as Parameters<typeof catchApiError>[0]));
      },
      finally: () => {
        $dispatch(endLoading(loadingId));
      },
    };
  };
