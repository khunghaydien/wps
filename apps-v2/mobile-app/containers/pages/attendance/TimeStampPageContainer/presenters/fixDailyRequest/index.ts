import { Store } from 'redux';

import cancelApproval from './cancelApproval';
import cancelSubmitted from './cancelSubmitted';
import submit from './submit';
import { bind } from '@apps/attendance/libraries/Collection';

export default (store: Store) =>
  bind(
    {
      submit,
      cancelSubmitted,
      cancelApproval,
    },
    store
  );
