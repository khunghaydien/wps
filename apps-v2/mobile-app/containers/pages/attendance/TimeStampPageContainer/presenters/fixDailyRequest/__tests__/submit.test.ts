import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import uuid from 'uuid/v4';

import { REASON } from '@attendance/domain/models/Result';

import presenter from '../submit';
import { IOutputData } from '@attendance/domain/useCases/fixDailyRequest/ISubmitUseCase';

jest.mock('uuid/v4');

const mockStore = configureMockStore([thunk]);

beforeEach(() => {
  jest.clearAllMocks();

  let i = 1;
  (uuid as jest.Mock).mockImplementation(() => `${i++}`);
});

describe('complete()', () => {
  it.each([
    { result: true },
    { result: false, reason: REASON.EXISTED_INVALID_REQUEST },
    { result: false, reason: REASON.EXISTED_SUBMITTING_REQUEST },
  ] as IOutputData[])('should do %s.', (output) => {
    // Arrange
    const store = mockStore({});

    // Act
    presenter(store)().complete(output);

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
  it(`should do with sending.`, () => {
    // Arrange
    const store = mockStore({
      attendance: {
        timeStamp: {
          ui: {
            sending: true,
          },
        },
      },
    });

    // Act
    presenter(store)().complete({ result: true, value: undefined });

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('error()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore({});

    // Act
    presenter(store)().error({
      message: 'Error Test',
    });

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('start()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore({});

    // Act
    presenter(store)().start();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('finally()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore({});

    // Act
    presenter(store)().finally();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});
