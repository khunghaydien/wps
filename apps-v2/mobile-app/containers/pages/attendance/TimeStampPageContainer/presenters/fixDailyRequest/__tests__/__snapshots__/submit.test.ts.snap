// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complete() should do { result: false, reason: 'existedInvalidRequest' }. 1`] = `
Array [
  Object {
    "payload": [Error: Error - There are requests that were rejected or canceled during the period.],
    "type": "CATCH_BUSINESS_ERROR",
  },
]
`;

exports[`complete() should do { result: false, reason: 'existedSubmittingRequest' }. 1`] = `
Array [
  Object {
    "payload": [Error: Error - There is a request waiting for approval during the period.],
    "type": "CATCH_BUSINESS_ERROR",
  },
]
`;

exports[`complete() should do { result: true }. 1`] = `
Array [
  Object {
    "payload": Object {
      "message": "Fix daily request is requested",
      "options": undefined,
      "variant": "success",
    },
    "type": "COMMONS/MODULES/TOAST/SHOW",
  },
  Object {
    "type": "ATTENDANCE/TIME_STAMP/UI/CLEAR_COMMENT",
  },
]
`;

exports[`complete() should do with sending. 1`] = `
Array [
  Object {
    "type": "ATTENDANCE/TIME_STAMP/UI/CLEAR_COMMENT",
  },
]
`;

exports[`error() should do. 1`] = `
Array [
  Object {
    "payload": Object {
      "errorCode": undefined,
      "isContinuable": true,
      "isFunctionCantUseError": false,
      "message": "APIエラー - Error Test",
      "problem": "Error Test",
      "stackTrace": "",
      "type": "APIエラー",
    },
    "type": "CATCH_API_ERROR",
  },
]
`;

exports[`finally() should do. 1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "END_LOADING",
  },
]
`;

exports[`start() should do. 1`] = `
Array [
  Object {
    "payload": "1",
    "type": "START_LOADING",
  },
]
`;
